tests/test_otel.py

```diff
@@ -0,0 +1,349 @@
+        }
+
+        location /context-inject {
+            otel_trace_context inject;
+            proxy_pass http://127.0.0.1:18080/notrace;
+        }
+
+        location /context-propagate {
+            otel_trace_context propagate;
+            proxy_pass http://127.0.0.1:18080/notrace;
+        }
+
+        location /notrace {
+            otel_trace off;
+            add_header "X-Otel-Trace-Id" $otel_trace_id;
+            add_header "X-Otel-Span-Id" $otel_span_id;
+            add_header "X-Otel-Parent-Id" $otel_parent_id;
+            add_header "X-Otel-Parent-Sampled" $otel_parent_sampled;
+            add_header "X-Otel-Traceparent" $http_traceparent;
+            add_header "X-Otel-Tracestate" $http_tracestate;
+            return 204;
+        }
+    }
+}
+
+"""
+
+(trace_id, span_id) = ("0af7651916cd43dd8448eb211c80319c", "b9c7c989f97918e1")
+context = {
+    "Traceparent": f"00-{trace_id}-{span_id}-01",
+    "Tracestate": "congo=ucfJifl5GOE,rojo=00f067aa0ba902b7",
+}
+
+
+def decode_id(span, value):
+    if value in ["trace_id", "span_id"]:
+        return binascii.hexlify(getattr(span, value)).decode("utf-8")
+    return value
+
+
+def get_attr(span, name):
+    for value in (a.value for a in span.attributes if a.key == name):
+        return getattr(value, value.WhichOneof("value"))
+
+
+def get_http09(url, logger):
+    def do_get(sock, path):
+        http_send = f"GET {path}\n".encode()
+        logger.debug(f"{http_send=}")
+        sock.sendall(http_send)
+        http_recv = sock.recv(1024)
+        logger.debug(f"{http_recv=}")
+        return http_recv.decode("utf-8")
+
+    p = urllib.parse.urlparse(url)
+    with socket.create_connection((p.hostname, p.port)) as sock:
+        recv = do_get(sock, p.path)
+    return recv
+
+
+@pytest.fixture
+def client(nginx):
+    urllib3.disable_warnings(urllib3.exceptions.InsecureRequestWarning)
+    with niquests.Session(multiplexed=True) as s:
+        yield s
+
+
+def test_response_http09(trace_service, nginx, logger):
+    assert get_http09("http://127.0.0.1:18080/ok", logger) == "OK"
+
+    span = trace_service.get_span()
+
+    assert span.name == "ok_location"
+
+
+@pytest.mark.parametrize("http_ver", ["1.1", "2.0", "3.0"])
+@pytest.mark.parametrize(
+    ("path", "status"),
+    [("/ok", 200), ("/err", 500), ("", 403)],
+    ids=["ok", "500 error", "403 error"],
+)
+def test_response_and_span_attributes(
+    http_ver, path, status, trace_service, client
+):
+    scheme, port = ("http", 18080) if http_ver == "1.1" else ("https", 18443)
+    if http_ver == "3.0":
+        client.quic_cache_layer.add_domain("127.0.0.1", port)
+    r = client.get(f"{scheme}://127.0.0.1:{port}{path}", verify=False)
+
+    span = trace_service.get_span()
+
+    assert span.name == ("ok_location" if path == "/ok" else path)
+
+    # Default span attributes
+    assert get_attr(span, "http.method") == "GET"
+    # Some inconsistence between target and route metrics
```

Comment:
It's expected, as "403" request doesn't have explicit location. Not sure, it's worth having such case. This response code looks like an accident rather then planned test case.

Reply:
OK, removed.